stdlib/public/Concurrency/TaskStatus.cpp

```diff
@@ -404,21 +382,42 @@ static bool swift_task_hasTaskGroupStatusRecordImpl() {
 /**************************************************************************/
 
 // ==== Child tasks ------------------------------------------------------------
-SWIFT_CC(swift)
-static ChildTaskStatusRecord*
-swift_task_attachChildImpl(AsyncTask *child) {
-  void *allocation = malloc(sizeof(swift::ChildTaskStatusRecord));
-  auto record = new (allocation) swift::ChildTaskStatusRecord(child);
-  SWIFT_TASK_DEBUG_LOG("attach child task = %p, record = %p, to current task = %p",
-                       child, record, swift_task_getCurrent());
-  swift_task_addStatusRecord(record);
-  return record;
-}
 
+/* Called in the path of linking a child into a parent/group synchronously with
+ * the parent task.
+ *
+ * When called to link a child into a parent directly, this does not hold the
+ * parent's task status record lock. When called to link a child into a task
+ * group, this holds the parent's task status record lock.
+ */
```

Comment:
should the more extensive docs on this func be in the header, and here none? 

PS: Ok I followed through how the status update is actually performed and it seems solid; we don't take the lock in the asyncLet case because we'd retry in `swift_task_addStatusRecordWithChecks` if the CAS to set new status failed it seems, right?

Reply:
So we have 2 options for how we implement `swift_task_addStatusRecordWithChecks`:

1. `swift_task_addStatusRecordWithChecks` actually takes the lock, which means the record in the atomic state is the lock record. So if we're trying to add a new record to the list, we have to add it as the parent to the innermost record inside the list. Having to take a lock just to add a record feels overkill if we don't expect contention most of the time - having to create the lock record, take the lock and then do the linkage inside.

2. `swift_task_addStatusRecordWithChecks` does the CAS loop and doesn't take the lock so that the new record is set up as the innermost record.

I chose option (2) for how to implement swift_task_addStatusRecordWithChecks which is why this comment says what it does.

Reply:
Right, got it 👍 Thanks for explaining again; agreed that this approach makes sense 👍

Reply:
> should the more extensive docs on this func be in the header, and here none?

Maybe? I think I usually stick it in the implementation cause my tags always jump to that and so all the information you need to know about the function is in one place.